lang/java/reef-utils/src/main/java/org/apache/reef/util/MultiAsyncToSync.java

```
@@ -0,0 +1,105 @@
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Assists a class in presenting a synchronous interface that is implemented
+ * via asynchronous interfaces and events. When a method call is received
+ * by the interface, parameter values captured and asynchronous processing
+ * started, the caller is put to sleep by calling block() with the internal
+ * interface caller identifier. After all of the asynchronous processing
+ * is complete the caller is released with a call to release().
+ */
+public final class MultiAsyncToSync {
+  private static final Logger LOG = Logger.getLogger(MultiAsyncToSync.class.getName());
+
+  private final ArrayDeque<ConditionVariable> freeQueue = new ArrayDeque<>();
+  private final HashMap<Long, ConditionVariable> sleeperMap = new HashMap<>();
+  private final long timeoutPeriod;
+  private final TimeUnit timeoutUnits;
+
+  /**
+   * Initialize a multiple asynchronous to synchronous object with a specified timeout value.
+   * @param timeoutPeriod The length of time in units geven by the the timeoutUnits
+   *                      parameter before the condition automatically times out.
+   * @param timeoutUnits The unit of time for the timeoutPeriod parameter.
+   */
+  public MultiAsyncToSync(final long timeoutPeriod, final TimeUnit timeoutUnits) {
+    this.timeoutPeriod = timeoutPeriod;
+    this.timeoutUnits = timeoutUnits;
+  }
+
+  /**
+   * Put the caller to sleep on a specific release identifier.
+   * @param identifier The identifier required to awake the caller via the release() method.
+   * @return
+   */
+  public boolean block(final long identifier) {
+    ConditionVariable call;
+    synchronized (this) {
+      // Get an condition variable to block the calling thread.
+      if (sleeperMap.containsKey(identifier)) {
+        // This should never happen as message identifiers are unique.
+        LOG.log(Level.SEVERE, "Duplicate identifier in RPC map");
```

It looks like we need to throw an exception here.